### ISSUE TITLE
Analyzer should assume that external factories can initialize final fields

### ISSUE THREAD

**Issue by fishythefish** [dart-lang/sdk#42955]
See the failures on https://dart-review.googlesource.com/c/sdk/+/157161 for context.

Analyzer warns that `foo` is not initialized, even though the factory will initialize it.

```dart
@JS()                                                                                                                                                         
@anonymous                                                                          
class C {                                                                           
  final foo;                                                                        
                                                                                    
  external factory C({foo});                                                        
}
```

**Comment by sigmundch** [comment]
Since we don't allow declaring non-external fields in JS-interop classes, I wonder if this issue would be handled once we have external fields.

My understanding is that external fields are meant to be supported when opting in to null-safety (see https://github.com/dart-lang/sdk/issues/41940), at that point then I hope we wont need anything special to handle non-external late fields initialized by external constructors.

Do you agree?

**Comment by fishythefish** [comment]
It depends on the details of that proposal/implementation. If `external` constructors are not allowed to initialize non-`external` fields, then we'd also need to update all JS-interop code to add `external` to fields, but this issue would likely be resolved since the analyzer can't check initialization of `external` fields. Having `external` fields should allow us to make them `late` without implementing our own lowering.

However, if `external` constructors can still initialize non-`external` fields (as is the case now), then there is a bug. The analyzer can't know the details of the `external` constructor implementation, so it should not produce an error. FWIW, this seems to be an issue with factories in particular - if the `factory` keyword is removed, analyzer accepts the code.

**Comment by lrhn** [comment]
I think this class should work whether the factory is external or not.

There is no path where the class is *instantiated* and the field is not initialized, because there is no generative constructor at all, so the class is never instantiated.

There should only be an error if a *non-redirecting generative* constructor fails to initialize a field which requires initialization. Classes without generative constructors satisfy this vacuously.

**Comment by eernstg** [comment]
This matches the case that we discussed at the language team meeting yesterday, and we decided that code with null safety should emit a compile-time error because `foo` is uninitialized (as far as the compiler knows). The problem is that `C` could be used to derive a mixin.

Until external fields arrive, the workaround could be to declare `get foo;` (if `C` can be abstract), and otherwise to keep the `late` modifier.
